### PR TITLE
fix: _REPO_DIGESTS is incorrect

### PR DESCRIPTION
**Pull request by miki725**
- [x] Followed the steps in the contributor's guide: https://crashoverride.com/docs/other/contributing#filing-the-pull-request
- [x] PR title uses [semantic commit messages](https://nitayneeman.com/posts/understanding-semantic-commit-messages-using-git-and-angular/#fix)
- [x] Filled out the template to a useful degree
- [x] Updated `CHANGELOG.md` if necessary

## Issue

docker mutats image digests in some cases hence `_IMAGE_DIGEST` is not correct semantically.

also there could be multiple list manifests all containing the same image hence `_IMAGE_LIST_DIGEST` is also not correct

## Description

* we are now validating all repo digests as reported by `docker inspect` to verify which digest is list or image digest
* dropping `_IMAGE_DIGEST` and `_IMAGE_LIST_DIGEST`
* updating structure of `_REPO_TAGS` and `_REPO_DIGESTS`
* adding `_REPO_LIST_DIGESTS`

## Testing

```
➜ reset; make tests args="test_docker.py::test_remanifest test_docker.py::test_retagging --logs --pdb -x"
```

### CODE REVIEW

**Comment by miki725** [comment]
https://app.shortcut.com/crashoverride-1/story/3284/chalk-wrong-digest-fields

File: src/docker/ids.nim [hunk 1]
```diff
@@ -384,6 +387,9 @@ proc `$`*(self: DockerImage): string =
       "docker image is empty to be represented correctly"
     )
 
+proc exists*(self: DockerImage): bool =
```
[the remark this answers is not in the document]
Comment (miki725) [reply src/docker/ids.nim:390]: this is a method for `DockerImage` specifically. the `inspect` `exists()` is to check if the image exists locally and so it has a different signature

normally we do `!= nil` checks however as `DockerImage` internally is a tuple, by default it gets initialized to `("", "", "")` however that doesnt mean that image ref actually exists hence this method to check it has some value

my preference is not to repeat data type in the function signatures

for example its `"".startsWith()` vs `"".stringStartsWith()`, etc

File: src/docker/ids.nim [hunk 2]
```diff
@@ -405,34 +411,98 @@ proc `$`*(items: seq[DockerImage]): seq[string] =
...
+  for i in self.digests:
+    yield (self.repo, "", i)
+
+iterator listManifests*(self: DockerImageRepo): DockerImage =
+  for i in self.listDigests:
+    yield (self.repo, "", i)
+
+iterator tagImages*(self: DockerImageRepo): DockerImage =
+  for i in self.tags:
+    yield (self.repo, i, "")
```
[the remark this answers is not in the document]
Comment (miki725) [reply src/docker/ids.nim:435]: there is no `yield from` in nim afaik so have to manually loop over underlying iterator and yield it again. might be a way to do it with template/macro but one more line is probably not a big deal I think

File: src/docker/ids.nim [hunk 3]
```diff
@@ -405,34 +411,98 @@ proc `$`*(items: seq[DockerImage]): seq[string] =
...
+  if a.repo != b.repo:
+    raise newException(ValueError, "cannot merge image repos for different repos")
+  return DockerImageRepo(
+    repo       : b.repo,
+    digests    : b.digests     + a.digests,
+    listDigests: b.listDigests + a.listDigests,
+    tags       : b.tags        + a.tags,
+  )
+
+# ----------------------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (miki725) [reply src/docker/ids.nim:468]: `DockerImageRepo` is a collection of sets of that image digests, list digests and tags. in nim `+` is an operator for set unions so I prefer to leave semantics similar with similar operators

can rename if this is confusing

File: src/docker/ids.nim [hunk 4]
```diff
@@ -405,34 +411,98 @@ proc `$`*(items: seq[DockerImage]): seq[string] =
...
   return self.name
 
-proc nameRef*(self: DockerManifest): DockerImage =
+# ----------------------------------------------------------------------------
+
+proc asImage*(self: DockerManifest): DockerImage =
   return self.name.withDigest(self.digest)
 
+proc asImageRepo*(self: DockerManifest): DockerImageRepo =
+  if self.kind != image:
```
[the remark this answers is not in the document]
Comment (miki725) [reply src/docker/ids.nim:481]: its an enum

https://github.com/crashappsec/chalk/blob/ebfd60e14955cf6f33646dd3d2c71efba7be4fe0/src/chalk_common.nim#L312-L313

as `self.kind` is of the same type you can just refer to the specific enum value directly like that:

https://github.com/crashappsec/chalk/blob/ebfd60e14955cf6f33646dd3d2c71efba7be4fe0/src/chalk_common.nim#L324

File: src/util.nim [hunk 5]
```diff
@@ -631,3 +631,10 @@ proc update*(self: HttpHeaders, with: HttpHeaders): HttpHeaders =
   for k, v in with.pairs():
     self[k] = v
   return self
+
+proc `+`*[T](a, b: OrderedSet[T]): OrderedSet[T] =
+  result = initOrderedSet[T]()
+  for i in a:
+    result.incl(i)
+  for i in b:
+    result.incl(i)
```
[the remark this answers is not in the document]
Comment (miki725) [reply src/util.nim:640]: yeah was surprised as well. there is one for `HashSet` but nothing for ordered sets. we prolly dont care about order in this case too much but wanted to leave whatever order docker reports back to us

File: src/docker/ids.nim (hunk 3 above)
[the remark this answers is not in the document]
Comment (mbaltrusitis) [reply src/docker/ids.nim:468]: its not confusion i'm worried about and more so keeping in mind how we're navigating the code base: `grep`ing a `+` will have much lower cardinality.

**Review by mbaltrusitis — APPROVED** [review]

**Pull request merged**